estate/models/estate_property.py

```diff
@@ -0,0 +1,85 @@
+from odoo.tools import relativedelta
+
+class EstateProperty(models.Model):
+
+    def _get_default_date_availability(self):
+        return fields.Date.today() + relativedelta(months=3)
+
+    _name = "estate.property"
+    _description = "Real Estate Property"
+
+    name = fields.Char(string="Title", required=True)
+    description = fields.Text()
+    active = fields.Boolean(default=True)
+    state = fields.Selection([
+        ('new', 'New'),
+        ('offer_received', 'Offer Received'),
+        ('offer_accepted', 'Offer Accepted'),
+        ('sold', 'Sold'),
+        ('canceled', 'Canceled')
+    ], default='new', required=True, copy=False)
+
+    tag_ids = fields.Many2many('estate.property.tag', string="Tags")
+    property_type_id = fields.Many2one('estate.property.type', string="Property Type")
+    salesman_id = fields.Many2one('res.users', string="Salesman", default=lambda self: self.env.user)
+    buyer_id = fields.Many2one('res.partner', string="Buyer", copy=False)
+    offer_ids = fields.One2many('estate.property.offer', 'property_id', string="Offers")
+
+    postcode = fields.Char(string="Postcode")
+    date_availability = fields.Date(string="Available From", default=_get_default_date_availability, copy=False)
+    expected_price = fields.Float(string="Expected Price", required=True)
+    selling_price = fields.Float(string="Selling Price", readonly=True, copy=False)
+
+    bedrooms = fields.Integer(string="Bedrooms", default=2)
+    living_area = fields.Integer(string="Living Area (sqm)")
+    facades = fields.Integer(string="Facades")
+    garage = fields.Boolean(string="Garage")
+    garden = fields.Boolean(string="Garden")
+    garden_area = fields.Integer()
+    garden_orientation = fields.Selection([
+        ('north', 'North'),
+        ('south', 'South'),
+        ('east', 'East'),
+        ('west', 'West')
+    ])
+
+    total_area = fields.Integer(string="Total Area (sqm)", compute="_compute_total_area")
+    best_price = fields.Float(string="Best Offer", compute="_compute_best_price")
+
+    @api.depends('living_area', 'garden_area')
+    def _compute_total_area(self):
+        for record in self:
+            record.total_area = record.living_area + record.garden_area
+
+    @api.depends('offer_ids.price')
+    def _compute_best_price(self):
+        for record in self:
+            record.best_price = max(record.mapped('offer_ids.price'), default=0)
+
+    @api.onchange('garden')
+    def _onchange_garden(self):
+        if not self.garden:
+            self.garden_area = 0
+            self.garden_orientation = None
+        else:
+            self.garden_area = 10
+            self.garden_orientation = 'north'
+
+    def action_cancel(self):
+        for record in self:
+            if record.state == 'sold':
+                raise exceptions.UserError("Sold properties cannot be canceled!")
+            else:
+                record.state = 'canceled'
+        return True
```

Comment:
1. Your method was called using a button that only exists on the form view, it would not make any sense if the recordset was containing more than one record.
2. Make sure to leave the responsibility of CRUD actions to the ORM as often as possible to benefit from the implicit optimizations.
=> Note that `record.state = 'canceled` was completely fine tho
```suggestion
    def action_cancel(self):
        self.ensure_one()
        if self.state == 'sold':
            raise exceptions.UserError("Sold properties cannot be canceled!")
        
        return self.write({"state": "canceled"})
```

Reply:
I understand 2. but for 1. would it be better to keep the for loop if after we want to add a feature that allow user to select some offers in order to cancel them all at once ?
Or even call this method in the accept method to cancel any other offers ?

What do you think ?

Reply:
Avoid over-anticipation